File: Python/uniquetol.py
# %%
import numpy as np
from numba import jit

# %%
eps = lambda x: np.finfo(x).eps

def isapprox(
        x: float, y: float, *,
        atol: float = 1e-8,
        rtol: float = None,
):
    if atol < 0:
        raise ValueError("`atol` must be non-negative")
    if rtol is None:
        rtol = max(eps(x), eps(y))**0.5
    elif rtol < 0:
        raise ValueError("`rtol` must be non-negative")
    
    return abs(x - y) <= max(atol, rtol * max(abs(x), abs(y)))

# %%
@jit(nopython=True)
def findfirst_farfrom(c: float, vec: np.ndarray, *, atol: float, rtol: float):
    idx = None
    for i, x in enumerate(vec):
        if abs(x - c) > max(atol, rtol * max(abs(c), abs(x))):
            idx = i
            break
    return idx

# %%
def uniquetol(
        vec: np.ndarray, *,
        atol: float = 1e-8,
        rtol: float = None,
        return_indices: bool = False,
        return_inverse: bool = False,
        return_counts: bool = False,
        occurrence: str = 'highest',
):
    if atol < 0:
        raise ValueError("`atol` must be non-negative")
    if rtol is None:
        rtol = eps(vec.dtype)**0.5
    elif rtol < 0:
        raise ValueError("`rtol` must be non-negative")
    if occurrence not in {'highest', 'lowest'}:
        raise ValueError('`occurrence` must be either "highest" or "lowest"')
    
    n = len(vec)
    
    if n == 0:
        vec_unique = np.empty(0)
        return_indices and (indices_unique := np.empty(0))
        return_inverse and (inverse_unique := np.empty(0))
        return_counts and (counts_unique := np.empty(0))
    else:
        isclose = lambda x, y: isapprox(x, y, atol=atol, rtol=rtol)
        
        perm_sorted = np.argsort(vec)
        vec_sorted = vec[perm_sorted]
        indices_unique = []
        idx = -1
        idx_switch = 0
        
        while idx_switch is not None:
            idx += idx_switch + 1
            indices_unique.append(idx)
            # if (idx_new := idx + 1) >= n:
            #     break
            # else:
            #     print(idx_new)
            c = vec_sorted[idx]
            idx_switch = findfirst_farfrom(c, vec_sorted[(idx + 1):], atol=atol, rtol=rtol)
        
        indices_unique = np.array(indices_unique)
        
        if return_counts or return_inverse:
            counts_unique = np.diff(np.append(indices_unique, n))
            if return_inverse:
                inverse_unique = np.concatenate(
                    [np.repeat(idx, ct) for idx, ct in enumerate(counts_unique)]
                )
                inverse_unique = inverse_unique[np.argsort(perm_sorted)]
        
        if occurrence == 'highest':
            indices_unique = np.append(indices_unique[1:] - 1, n - 1)
        
        indices_unique = perm_sorted[indices_unique]
        vec_unique = vec[indices_unique]
    
    if return_indices or return_inverse or return_counts:
        output_temp = []
        return_indices and output_temp.append(indices_unique)
        return_inverse and output_temp.append(inverse_unique)
        return_counts and output_temp.append(counts_unique)
        output = tuple([vec_unique, *output_temp])
    else:
        output = vec_unique
    
    return output

File: C/uniquetol.c
//
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <math.h>

#include "isapprox.h"
#include "sortperm.h"
#include "uniquetol.h"

//
UniqueTolArray uniquetol_base(
    double *arr,
    int n,
    double atol,
    double rtol,
    char *occurrence
);

//
UniqueTolArray uniquetol_var(double *arr, int n, UniqueTolArgs in) {
    double atol_out = in.atol ? in.atol : 1e-8;
    double rtol_out = in.rtol ? in.rtol : sqrt(nextafter(1, 2) - 1);
    char *occurrence_out = in.occurrence ? in.occurrence : "highest";
    return uniquetol_base(arr, n, atol_out, rtol_out, occurrence_out);
}

//
UniqueTolArray uniquetol_base(
    double *arr,
    int n,
    double atol,
    double rtol,
    char *occurrence
) {
    int use_highest = strcmp(occurrence, "highest");
    int use_lowest = strcmp(occurrence, "lowest");
    
    if (use_highest != 0 && use_lowest != 0) {
        fprintf(stderr, "`occurrence` must be either `highest` or `lowest`");
        exit(1);
    }
    
    UniqueTolArray out;
    
    if (n == 0) {
        double arr_unique[0] = {};
        int indices_unique[0] = {};
        int inverse_unique[0] = {};
        int counts_unique[0] = {};
        
        out = (UniqueTolArray){
            .arr_unique = arr_unique,
            .indices_unique = indices_unique,
            .inverse_unique = inverse_unique,
            .counts_unique = counts_unique,
            .num_total = 0,
            .num_unique = 0
        };
    }
    else {
        int perm_sorted[n];
        double arr_sorted[n];
        int inverse_unique[n];
        int *indices_unique_temp = (int*)malloc(n * sizeof(int));
        int i = 0; int j = 1; int k = 0;
        sortperm(arr, perm_sorted, n);
        
        for (int i = 0; i < n; i++) {
            arr_sorted[i] = arr[perm_sorted[i]];
        }
        
        while (j < n) {
            double c = arr_sorted[i];
            bool is_close = true;
            
            while (is_close && j < n) {
                is_close = isapprox(c, arr_sorted[j], atol, rtol);
                j++;
            }
            
            i = j - 1;
            
            if (!is_close) {
                indices_unique_temp[k] = i;
                k++;
            }
        }
        
        double arr_unique[k];
        int counts_unique[k];
        int *indices_unique = (int*)realloc(indices_unique_temp, k * sizeof(int));
        counts_unique[k] = n - indices_unique[k];
        
        for (int i = 0; i < k - 1; i++) {
            counts_unique[i] = indices_unique[i + 1] - indices_unique[i];
        }
        
        for (int i = 0; i < k; i++) {
            for (int j = 0; j < counts_unique[i]; j++) {
                inverse_unique[indices_unique[i] + j] = i;
            }
        }
        
        if (use_highest == 0) {
            for (int i = 0; i < k - 1; i ++) {
                indices_unique[i] = indices_unique[i + 1] - 1;
            }
            
            indices_unique[k - 1] = n - 1;
        }
        
        for (int i = 0; i < k; i++) {
            indices_unique[i] = perm_sorted[indices_unique[i]];
        }
        
        for (int i = 0; i < k; i++) {
            arr_unique[i] = arr[indices_unique[i]];
        }
        
        out = (UniqueTolArray){
            .arr_unique = arr_unique,
            .indices_unique = indices_unique,
            .inverse_unique = inverse_unique,
            .counts_unique = counts_unique,
            .num_total = n,
            .num_unique = k
        };
    }
    
    return out;
}